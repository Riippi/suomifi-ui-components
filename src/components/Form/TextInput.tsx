import React, { Component, ReactNode, ChangeEvent, FocusEvent } from 'react';
import { Omit } from '../../utils/typescript';
import {
  HtmlLabel,
  HtmlLabelProps,
  HtmlInput,
  HtmlInputProps,
  HtmlDiv,
  HtmlDivProps,
  HtmlSpan,
} from '../../reset';
import { VisuallyHidden } from '../Visually-hidden/Visually-hidden';
import { Paragraph, ParagraphProps } from '../Paragraph/Paragraph';
import classnames from 'classnames';
import styled from 'styled-components';
import { disabledCursor } from '../utils/css';
import { idGenerator } from '../../utils/uuid';

const baseClassName = 'fi-text-input';
const disabledClassName = `${baseClassName}--disabled`;
const labelBaseClassName = `${baseClassName}_label`;
const inputBaseClassName = `${baseClassName}_input`;
const statusTextClassName = `${baseClassName}_statusText`;
const statusTextContainerClassName = `${statusTextClassName}_container`;
const hintTextClassName = `${baseClassName}_hintText`;

export interface TextInputLabelProps extends HtmlLabelProps {}

type Label = 'hidden' | 'visible';

type InputType = 'text' | 'email' | 'number' | 'password' | 'tel' | 'url';

type TextInputStatus = 'default' | 'error' | 'success';

export interface TextInputProps extends Omit<HtmlDivProps, 'type'> {
  /** Custom classname for the input to extend or customize */
  className?: string;
  /** Disable input usage */
  disabled?: boolean;
  /** Event handler to execute when clicked */
  onClick?: () => void;
  /** Pass custom props to label container */
  labelProps?: TextInputLabelProps;
  /** Pass custom props to Label text element */
  labelTextProps?: ParagraphProps;
  /** To execute on input text change */
  onChange?: (event: ChangeEvent<HTMLInputElement>) => void;
  /** To execute on input text onBlur */
  onBlur?: (event: FocusEvent<HTMLInputElement>) => void;
  /** Label */
  labelText: string;
  /** Hide or show label. Label element is always present, but can be visually hidden.
   * @default visible
   */
  labelMode?: Label;
  /** Default value */
  defaultValue?: string;
  /** Placeholder text for input. Use only as visual aid, not for instructions. */
  visualPlaceholder?: string;
  /** Input container div to define custom styling */
  inputContainerProps?: HtmlDivProps;
  /** Input element props */
  inputProps?: HtmlInputProps;
  /** A custom element to be passed to the component. Will be rendered after the input */
  children?: ReactNode;
  /** Hint text to be shown below the component */
  hintText?: string;
  /**
   * 'default' | 'error' | 'success'
   * @default default
   */
  status?: TextInputStatus;
  /** Status text to be shown below the component and hint text. Use e.g. for validation error */
  statusText?: string;
  /** 'text' | 'email' | 'number' | 'password' | 'tel' | 'url'
   * @default text
   */
  type?: InputType;
  /** Input name */
  name?: string;
  /** Input id */
  id?: string;
  /** Set components width to 100% */
  fullWidth?: boolean;
}

class BaseTextInput extends Component<TextInputProps> {
  render() {
    const {
      disabled,
      labelText,
      onBlur,
      onClick,
      onChange,
      labelMode,
      labelProps,
      labelTextProps,
      defaultValue,
      inputContainerProps,
      inputProps,
      children,
      status,
      statusText,
      hintText,
      visualPlaceholder,
      id: propId,
      type = 'text',
      name,
      fullWidth,
      ...passProps
    } = this.props;

    const hideLabel = labelMode === 'hidden';
    const generatedStatusTextId = `${idGenerator(propId)}-statusText`;
    const generatedHintTextId = `${idGenerator(propId)}-hintText`;
    const inputElementProps = {
      ...inputProps,
      className: classnames(inputBaseClassName, inputProps?.className),
      id: propId,
      name,
      type,
      disabled,
      defaultValue,
      placeholder: visualPlaceholder,
      ...{ 'aria-invalid': status === 'error' },
      onBlur,
      onClick,
      onChange,
    };

    return (
      <HtmlDiv {...passProps}>
        <HtmlLabel
          {...labelProps}
          className={classnames(labelBaseClassName, {
            [disabledClassName]: !!disabled,
          })}
        >
          {hideLabel ? (
            <VisuallyHidden>{labelText}</VisuallyHidden>
          ) : (
            <Paragraph {...labelTextProps}>{labelText}</Paragraph>
          )}
          {hintText && (
            <Paragraph className={hintTextClassName} id={generatedHintTextId}>
              {hintText}
            </Paragraph>
          )}
          <HtmlDiv className={statusTextContainerClassName}>
            <HtmlDiv {...inputContainerProps}>
              <HtmlInput {...inputElementProps} />
              {children}
            </HtmlDiv>
            {statusText && (
              <HtmlSpan
                className={statusTextClassName}
                id={generatedStatusTextId}
              >
                {statusText}
              </HtmlSpan>
            )}
          </HtmlDiv>
        </HtmlLabel>
      </HtmlDiv>
    );
  }
}

export const StyledBaseTextInput = styled((props: TextInputProps) => (
  <BaseTextInput {...props} />
))`
  &.${disabledClassName} {
    ${disabledCursor}
  }
`;

export class TextInput extends Component<TextInputProps> {
  render() {
    return <StyledBaseTextInput {...this.props} />;
  }
}
