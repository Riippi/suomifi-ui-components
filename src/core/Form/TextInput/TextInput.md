```js
import { TextInput } from 'suomifi-ui-components';

<TextInput
  onBlur={event => console.log(event.target.value)}
  labelText="Test TextInput"
/>;
```

```js
import { TextInput } from 'suomifi-ui-components';

<>
  <TextInput
    onBlur={event => console.log(event.target.value)}
    labelMode="hidden"
    labelText="Test with hidden label"
    defaultValue="Test with hidden label"
  />
  <TextInput.error
    labelMode="hidden"
    labelText="Error with hidden label"
    defaultValue="Error with hidden label"
  />
  <TextInput.success
    labelMode="hidden"
    labelText="Success with hidden label"
    defaultValue="Success with hidden label"
  />
</>;
```
