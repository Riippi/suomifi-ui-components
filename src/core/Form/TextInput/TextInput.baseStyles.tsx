import { css } from 'styled-components';
import { suomifiTheme } from '../../theme';
import { TextInputProps } from './TextInput';
import { input, inputContainer } from '../../theme/reset';

export const baseStyles = ({ theme = suomifiTheme }: TextInputProps) => css`
  & .fi-text-input_label-p {
    margin-bottom: ${theme.spacing.m};
  }

  & .fi-text-input_container {
    ${inputContainer(theme)}
  }

  & .fi-text-input_input {
    ${input(theme)}
    background-color: ${theme.colors.whiteBase};
  }

  &.fi-text-input--error {
    & .fi-text-input_input {
      color: ${theme.colors.alertBase};
      border-color: ${theme.colors.alertBase};
    }
  }
  &.fi-text-input--success {
    & .fi-text-input_input {
      color: ${theme.colors.successBase};
      border-color: ${theme.colors.successBase};
    }
  }
`;
