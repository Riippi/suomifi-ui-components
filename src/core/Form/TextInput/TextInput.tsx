import React, { Component } from 'react';
import styled from '@emotion/styled';
import { withDefaultTheme } from '../../theme/utils';
import { ThemeComponent } from '../../theme';
import { baseStyles } from './TextInput.baseStyles';
import {
  TextInput as CompTextInput,
  TextInputProps as CompTextInputProps,
} from '../../../components/Form/TextInput';
import classnames from 'classnames';

const baseClassName = 'fi-text-input';
const labelParagraphClassName = `${baseClassName}_label-p`;
const inputContainerClassName = `${baseClassName}_container`;
const errorClassName = `${baseClassName}--error`;
const successClassName = `${baseClassName}--success`;

type TextInputVariant = 'default' | 'error' | 'success';
export interface TextInputProps extends CompTextInputProps, ThemeComponent {
  variant?: TextInputVariant;
}

const StyledTextInput = styled(
  ({
    theme,
    variant,
    className,
    labelTextProps = { className: undefined },
    inputContainerProps = { className: undefined },
    ...passProps
  }: TextInputProps) => {
    return (
      <CompTextInput
        {...passProps}
        labelTextProps={{
          ...labelTextProps,
          className: classnames(
            labelTextProps.className,
            labelParagraphClassName,
          ),
        }}
        inputContainerProps={{
          ...inputContainerProps,
          className: classnames(
            inputContainerProps.className,
            inputContainerClassName,
          ),
        }}
        className={classnames(className, {
          [errorClassName]: variant === 'error',
          [successClassName]: variant === 'success',
        })}
      />
    );
  },
)`
  ${props => baseStyles(props)}
`;

/**
 * <i class="semantics" />
 * Use for user inputting text
 */
export class TextInput extends Component<TextInputProps> {
  static error = (props: TextInputProps) => (
    <StyledTextInput {...withDefaultTheme(props)} variant="error" />
  );

  static success = (props: TextInputProps) => (
    <StyledTextInput {...withDefaultTheme(props)} variant="success" />
  );

  render() {
    const { ...passProps } = withDefaultTheme(this.props);

    return <StyledTextInput {...passProps} />;
  }
}
