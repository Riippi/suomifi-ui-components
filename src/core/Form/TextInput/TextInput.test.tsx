import React from 'react';
import { render } from 'react-testing-library';
import { axeTest } from '../../../utils/test/axe';

import { TextInput } from './TextInput';

const TestTextInput = (
  <TextInput labelText="Test input" data-testid="textinput" />
);

test('calling render with the same component on the same container does not remount', () => {
  const buttonRendered = render(TestTextInput);
  const { container } = buttonRendered;
  expect(container.firstChild).toMatchSnapshot();
});

test('should not have basic accessibility issues', axeTest(TestTextInput));
