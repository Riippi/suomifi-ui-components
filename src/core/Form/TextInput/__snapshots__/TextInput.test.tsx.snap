// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling render with the same component on the same container does not remount 1`] = `
.emotion-3 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  max-width: 100%;
}

.emotion-3 .fi-text-input_label-p {
  margin-bottom: 16px;
}

.emotion-3 .fi-text-input_container > input:focus {
  outline-color: hsl(23,82%,53%);
  outline-width: 4px;
  outline-offset: 2px;
}

.emotion-3 .fi-text-input_container:focus-within {
  outline: 0;
  position: relative;
}

.emotion-3 .fi-text-input_container:focus-within:after {
  content: '';
  position: absolute;
  top: -4px;
  right: -4px;
  bottom: -4px;
  left: -4px;
  border-radius: 4px;
  background-color: transparent;
  border: 1px solid hsl(23,82%,53%);
  box-sizing: border-box;
  box-shadow: 0 0 10px 0 hsl(23,82%,53%);
  z-index: 9999;
}

.emotion-3 .fi-text-input_container:focus-within:not(:focus-visible):after {
  content: none;
}

.emotion-3 .fi-text-input_container:focus-within > input:focus {
  outline: none;
}

.emotion-3 .fi-text-input_input {
  color: hsl(0,0%,16%);
  font-family: 'Source Sans Pro','Helvetica Neue',Arial,sans-serif;
  font-size: 16px;
  font-weight: 400;
  line-height: 1.5;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  min-width: 245px;
  max-width: 100%;
  padding: 8px 16px;
  border: 1px solid hsl(202,7%,67%);
  border-radius: 2px;
  line-height: 1;
  background-color: hsl(0,0%,100%);
}

.emotion-3.fi-text-input--error .fi-text-input_input {
  color: hsl(3,59%,48%);
  border-color: hsl(3,59%,48%);
}

.emotion-3.fi-text-input--success .fi-text-input_input {
  color: hsl(166,90%,36%);
  border-color: hsl(166,90%,36%);
}

.emotion-0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.emotion-1 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  overflow: visible;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
}

.emotion-1::-webkit-input-placeholder {
  color: inherit;
  opacity: 0.54;
}

<label
  class="fi-text-input_label emotion-3"
>
  <p
    class="fi-paragraph fi-text-input_label-p emotion-0"
  >
    Test input
  </p>
  <div
    class="fi-text-input_container emotion-0"
  >
    <input
      class="fi-text-input_input emotion-1"
      data-testid="textinput"
      type="text"
    />
  </div>
</label>
`;
